### PR TITLE
Fix: python slices of splits used incorrect offsets.

### PR DESCRIPTION
**Pull request by kmapb**
32-bit and 64-bit offset logic assumed a `Strs` produced from by slicing a previous `Strs` started at offset 0. `str.split().slice([1:])`, for instance, produced impossible end offsets. This was the cause of some SEGVs and negative-length Python string constructions.

Deduplicate the 32-bit and 64-bit logic, and fix the bug in the single copy of the logic.

Test for 32-bit and 64-bit indices. Previous version of `lib.c` caused this test to SEGV or produce runtime errors about constructing Python strings of negative length depending on memory conditions.

### CODE REVIEW

File: python/lib.c [hunk 1]
```diff
@@ -732,44 +732,37 @@ static PyObject *Strs_subscript(Strs *self, PyObject *key) {
...
 
-            size_t first_length;
-            str_at_offset_consecutive_32bit(self, start, count, &to->parent, &to->start, &first_length);
-            uint32_t first_offset = to->start - from->start;
-            to->end_offsets = malloc(sizeof(uint32_t) * to->count);
-            if (to->end_offsets == NULL && PyErr_NoMemory()) {
-                Py_XDECREF(self_slice);
-                return NULL;
-            }
-            for (size_t i = 0; i != to->count; ++i) to->end_offsets[i] = from->end_offsets[i] - first_offset;
```
Comment (kmapb) [inline python/lib.c:750]: This was one copy of the actual bug; `from_offsets` should only run from `0` to `length` if `start` was 0.

File: python/lib.c [hunk 2]
```diff
@@ -732,44 +732,37 @@ static PyObject *Strs_subscript(Strs *self, PyObject *key) {
...
+    to->parent = from->parent;                                                                                \
+    size_t first_length;                                                                                      \
+    str_at_offset_consecutive_##type(self, start, count, &to->parent, &to->start, &first_length);             \
+    index_t first_offset = to->start - from->start;                                                           \
+    to->end_offsets = malloc(sizeof(index_t) * to->count);                                                    \
+    if (to->end_offsets == NULL && PyErr_NoMemory()) {                                                        \
+        Py_XDECREF(self_slice);                                                                               \
+        return NULL;                                                                                          \
+    }                                                                                                         \
+    for (size_t i = 0; i != to->count; ++i) to->end_offsets[i] = from->end_offsets[i + start] - first_offset; \
```
Comment (kmapb) [inline python/lib.c:755]: This is the unified fix.

File: python/lib.c [hunk 3]
```diff
@@ -732,44 +732,37 @@ static PyObject *Strs_subscript(Strs *self, PyObject *key) {
...
-
-            size_t first_length;
-            str_at_offset_consecutive_64bit(self, start, count, &to->parent, &to->start, &first_length);
-            uint64_t first_offset = to->start - from->start;
-            to->end_offsets = malloc(sizeof(uint64_t) * to->count);
-            if (to->end_offsets == NULL && PyErr_NoMemory()) {
-                Py_XDECREF(self_slice);
-                return NULL;
-            }
-            for (size_t i = 0; i != to->count; ++i) to->end_offsets[i] = from->end_offsets[i] - first_offset;
```
Comment (kmapb) [inline python/lib.c:769]: This was the other copy of the bug.

**Review by kmapb — COMMENTED** [review]
leaving some breadcrumbs for real reviewers...

**Comment by kmapb** [comment]
Against current main-dev

**Comment by ashvardanian** [comment]
Didn't realize the PR contains a patch until today, sorry, will merge soon 🤗

**Review by ashvardanian — APPROVED** [review]

**Pull request merged**